libtenzir/src/metric_handler.cpp

```diff
@@ -0,0 +1,47 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2024 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "tenzir/metric_handler.hpp"
+
+#include "tenzir/pipeline.hpp"
+
+#include <caf/send.hpp>
+
+namespace tenzir {
+
+metric_handler::metric_handler(
+  detail::weak_handle<metrics_receiver_actor> new_receiver,
+  uint64_t operator_index, type metric_type)
+  : receiver_{std::move(new_receiver)}, op_index_{operator_index} {
+  const auto* metric_record_type = caf::get_if<record_type>(&metric_type);
+  TENZIR_ASSERT(metric_record_type);
+  if (metric_record_type->num_fields() == 0) {
+    return;
+  }
+  auto params_type = metric_record_type->transform({{
+    {0},
+    record_type::insert_before({
+      {"timestamp", time_type{}},
+      {"operator_id", uint64_type{}},
+    }),
+  }});
+  metric_type_ = type{metric_type.name(), *params_type, {{"internal", ""}}};
+}
+
+auto metric_handler::emit(record&& r) -> void {
+  // Explicitly create a data-from-time cast here to support macOS builds.
+  r["timestamp"] = data{time{time::clock::now()}};
+  r["operator_id"] = op_index_;
+  caf::anon_send(receiver_.lock(), metric_type_, std::move(r));
```

Comment:
There's one problem with this approach: Sending the type along every single time is cheap when the operator is in the same process as the diagnostic handler, and absurdly expensive when it isn't.

Do you wanna pair sometime today to make this cheaper and push this PR over the finish line?

Reply:
OK; after I prepared my other PR.